Module4/Assignment4/Assignment4.cpp

```diff
@@ -7,33 +7,187 @@
 #include <stdio.h>
 #include <math.h>
 #include <iostream>
+#include <string.h>
 using namespace std;
 
+bool isSorted(int *numbers, int size)
+{
+    if (size <= 1)
+    {
+        return true;
+    }
+    int previousElement = numbers[0];
+    for (int i = 1; i < size; i++)
+    {
+        if (previousElement > numbers[i])
+        {
+            return false;
+        }
+        previousElement = numbers[i];
+    }
+    return true;
+}
+void printArray(int *numbers, int size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        cout << numbers[i] << ", ";
+    }
+    cout << endl;
+}
+
+const int MAX_SIZE = 128;
+int merge(int theArray[], int first, int mid, int last)
+{
+    int accessCount = 0;
+    int tempArray[MAX_SIZE];
+    memset(tempArray, 0xdeadbeef, sizeof(tempArray));
+    int first1 = first;   // begin first subarray
+    int last1 = mid;      // end first subarray
+    int first2 = mid + 1; // begin second subarray
+    int last2 = last;     // end second subarray
+    int index = first1;
+    while ((first1 <= last1) && (first2 <= last2))
+    {
+        accessCount += 2;
+        if (theArray[first1] <= theArray[first2])
+        {
+            tempArray[index] = theArray[first1];
+            accessCount++;
+            first1++;
+        }
+        else
+        {
+            tempArray[index] = theArray[first2];
+            accessCount++;
+            first2++;
+        } // end if
+        index++;
+    } // end while
+    while (first1 <= last1)
+    {
+        tempArray[index] = theArray[first1];
+        accessCount++;
+        first1++;
+        index++;
+    } // end while
+    while (first2 <= last2)
+    {
+        tempArray[index] = theArray[first2];
+        accessCount++;
+        first2++;
+        index++;
+    }
+    for (index = first; index <= last; index++)
+    {
+        accessCount++;
+        theArray[index] = tempArray[index];
+    }
+    return accessCount;
+} // end merge
+
+int mergeSort(int theArray[], int first, int last)
+{
+    int accessCount = 0;
+    if (first < last)
+    {
+        int mid = first + ((last - first) / 2); // midpoint index
+        accessCount += mergeSort(theArray, first, mid);
+        accessCount += mergeSort(theArray, mid + 1, last);
+        accessCount += merge(theArray, first, mid, last);
+
+    } // end if
+    return accessCount;
+}
+
 int mergeSort(int *numbers, int size)
 {
-    return -1;
+    // cout << "before: " << endl;
+    // printArray(numbers, size);
+    int accessCount = size > 0 ? mergeSort(numbers, 0, size - 1) : 0;
+    // cout << "after: " << endl;
+    // printArray(numbers, size);
+    return accessCount;
 }
+
 int bubbleSort(int *numbers, int size)
 {
-    return -1;
+    int accessCount = 0;
+    for (int i = 0; i < size - 1; i++)
+    {
+        for (int j = 0; j < size - i - 1; j++)
+        {
+            if (numbers[j] > numbers[j + 1])
```

Comment:
Since we access `numbers[j + 1]` three times in this function, we should make a local variable for it at the top of the `for` loop: ```
int nextIndex = j + 1;
int nextNumber = numbers[nextIndex];
`

then use these values in the rest of the code as appropriate.

(this kind of commentary is called "code golf" but it's good to get comfortable with it)